torchtune/models/clip/_component_builders.py

```diff
@@ -140,3 +152,334 @@ def clip_mlp(in_dim: int, out_dim: int, hidden_dim: int, activation: nn.Module,
+            If provided, it will return the intermediate results of the transformer layers
+            before they go through a next layer. For example, ``out_indices=[0,3]`` will
+            return the tokens before they go through the first and fourth layers.
+        output_cls_projection (bool): If True, only the CLS token projection will be outputted,
+            instead of all tokens. Defaults to False.
+        max_num_tiles (int): The maximum number of tiles that can be processed. This is used to
+            determine the size of the positional embeddings.
+        in_channels (int): The number of image input channels.
+        intermediate_act (torch.nn.Module): The activation function used in the intermediate layers in the transformer encoder.
+        lora_rank (int): rank of each low-rank approximation
+        lora_alpha (float): scaling factor for the low-rank approximation
+        lora_dropout (float): LoRA dropout probability. Default: 0.0
+        use_dora (bool): Whether to use DoRA layers instead of LoRA layers. Default is ``False``.
+        quantize_base: (bool): Whether to quantize base model weights or not. Only applied to base
+            weights within linear layers LoRA is applied to. The final output linear projection is not
+            supported for quantization currently.
+        
+
+    Returns:
+        VisionTransformer: Instantiation of VisionTransformer model.
+    """
+    assert embed_dim % num_heads == 0, "embed_dim must be divisible by num_heads"
+
+    # TODO: add support for quantizing and LoRA for the final output projection
+    cls_projection = CLSProjection(embed_dim=embed_dim, cls_output_dim=cls_output_dim) if output_cls_projection else None
+
+    # transformer layer
+    self_attn = lora_clip_attention(
+            lora_modules=lora_modules,
+            embed_dim=embed_dim,
+            num_heads=num_heads,
+            num_kv_heads=num_heads,
+            head_dim=embed_dim // num_heads,
+            attn_dropout=0.0,
+            lora_rank=lora_rank,
+            lora_alpha=lora_alpha,
+            lora_dropout=lora_dropout,
+            use_dora=use_dora,
+            quantize_base=quantize_base,
+    )
+    if apply_lora_to_mlp:
+        mlp = lora_clip_mlp(
+                in_dim=embed_dim,
+                hidden_dim=4 * embed_dim,
+                out_dim=embed_dim,
+                activation=activation(),
+                lora_rank=lora_rank,
+                lora_alpha=lora_alpha,
+                quantize_base=quantize_base,
+                lora_dropout=lora_dropout,
+                use_dora=use_dora,
+            )
+    else:
+        mlp = clip_mlp(
+            in_dim=embed_dim,
+            hidden_dim=4 * embed_dim,
+            out_dim=embed_dim,
+            activation=activation(),
+            quantize_base=quantize_base,
+        )
+    transformer_layer = TransformerSelfAttentionLayer(
```

Comment:
do you mind using ModuleList like in llama 3.1, so we can avoid using _copy?

Reply:
I'd rather leave this for a future depreciation of the layer option

Reply:
does it use rope embeddings? If so, it will consume more memory. Either way works for me

Reply:
Do it @pbontrager